File: src/lexer/source_file.cpp
#include "lexer/source_file.h"
#include "lexer/source_line.h"

SourceFile::SourceFile(const str& filePath)
: file(std::ifstream(filePath)), path(filePath), line(nullptr) {
    readLine();
}

SourceFile::~SourceFile() {
    file.close();
    for(SourceLine* line : lines) delete line;
}

void SourceFile::readLine() {
    nat num = (line == nullptr) ? 1 : line->num + 1;
    lines.push_back(line = new SourceLine(num));

    char c;

    while(file.good()) {
        c = file.get();
        if(c == EOF) return;
        line->string += c;

        if(c == '\n') return;
    }
}

char SourceFile::next() {
    str string = line->unreadString();
    return (string != "") ? string[0] : '\0';
}

str SourceFile::take(nat num, const str& these, const str& until) {
    str string;

    if(line == nullptr or num == 0) return string;

    for(char c : line->unreadString()) {
        if(in(c, until) or (these != "" and not in(c, these)))
            return string; 

        string += c;
        line->increment();

        if(string.length() == num) break;
    }

    if(line->unreadString() == "" and line->string != "") readLine();
    return string;
}

File: app/src/main/kotlin/sea/lexer/SourceFile.kt
package sea.lexer

import java.io.File
import sea.lexer.SourceLine

data class SourceFile(val path: String) {
    val lines: ArrayList<SourceLine> = ArrayList()

    init {
        File(path).bufferedReader().useLines { fileLines ->
            for((i, line) in fileLines.withIndex()) {
                lines.add(SourceLine(i + 1, "$line\n"))
            }
        }

        lines.add(SourceLine(lines.size + 1, ""))
    }

    var line: SourceLine = lines[0]

    val atEnd: Boolean
        get() = line.num > lines.size - 1

    val next: Char
        get() = if(line.unreadString != "") line.unreadString[0] else '\u0000'

    fun iterateLine() {
        if(!atEnd) line = lines[line.num]
    }

    fun take(num: Int = -1, these: String = "", until: String = ""): String {
        if(atEnd || num == 0) return ""
        var string = ""

        for(c in line.unreadString) {
            if(c in until || (these != "" && !(c in these))) return string

            string += c
            line.increment()

            if(string.length == num) break
        }

        if(line.unreadString == "") iterateLine()
        return string
    } 
}
